### PR TITLE
Feature/smart get lookup key field

### PR DESCRIPTION
**Pull request by prescod**
Behaviour change:

 * lookups do not anymore default to using snake_case, although they will look for snake_case in a dataset.

 * a few code paths which previously did not use our official mapping parser, now do. Which means that they will give error messages for misplaced, misnamed or missing keys

Internal renamings/refactorings:

 * some classes used "self.mappings" and others "self.mapping". Now they all use "self.mapping".

 * get_lookup_key_field is now a method which takes 0 arguments, not a function which takes 2. The old function remains temporarily for backwards compatibility. (primarily with snowfakery)

### CODE REVIEW

File: cumulusci/tasks/bulkdata/mapping_parser.py [hunk 1]
```diff
@@ -19,6 +21,28 @@ class MappingLookup(CCIDictModel):
...
+
+        # CCI used snake_case until mid-2020.
+        # At some point this code could probably be simplified.
+        snake_cased_guesses = list(map(convert_to_snake_case, guesses))
+        for guess in guesses + snake_cased_guesses:
+            if hasattr(model, guess):
+                return guess
+        raise KeyError(
+            f"Could not find a key field for {self.name}.\n"
+            + f"Tried {', '.join(snake_cased_guesses)}"
```
Comment (davisagli) [inline cumulusci/tasks/bulkdata/mapping_parser.py:44]:
```suggestion
            + f"Tried {', '.join(guesses)}"
```

File: cumulusci/tasks/bulkdata/mapping_parser.py [hunk 2]
```diff
@@ -19,6 +21,28 @@ class MappingLookup(CCIDictModel):
     join_field: Optional[str] = None
     after: Optional[str] = None
     aliased_table: Optional[str] = None
+    name: Optional[str] = None  # populated by parent
+
+    def get_lookup_key_field(self, model=None):
```
Comment (davisagli) [inline cumulusci/tasks/bulkdata/mapping_parser.py:26]: docstring please

File: cumulusci/tasks/bulkdata/mapping_parser.py [hunk 3]
```diff
@@ -50,6 +75,12 @@ def oid_as_pk_is_deprecated(cls, v):
         )
         return v
 
+    @root_validator  # not really a validator, more like a post-processor
+    def fixup_lookup_names(cls, v):
```
Comment (davisagli) [inline cumulusci/tasks/bulkdata/mapping_parser.py:79]: docstring please

File: cumulusci/tasks/bulkdata/utils.py [hunk 4]
```diff
@@ -9,11 +10,16 @@
...
 from cumulusci.core.exceptions import BulkDataException
-from cumulusci.utils import convert_to_snake_case
 
 
-def get_lookup_key_field(lookup, sf_field):
-    return lookup.get("key_field") or convert_to_snake_case(sf_field)
+def get_lookup_key_field(lookup, model=None):
+    # This should be removed in late 2020.
+    # It just provides some temporary backwards-compatibility
+    # for Snowfakery
```
Comment (davisagli) [inline cumulusci/tasks/bulkdata/utils.py:18]: What do you mean about backwards-compatibility for Snowfakery? This function is not used in Snowfakery.

Comment (prescod) [reply cumulusci/tasks/bulkdata/utils.py:18]: I forgot that it is only called indirectly. You think it's a safe bet that nobody is using it externally?

File: cumulusci/tasks/bulkdata/mapping_parser.py (hunk 3 above)
Comment (prescod) [reply cumulusci/tasks/bulkdata/mapping_parser.py:81]: 3441a8c8

File: cumulusci/tasks/bulkdata/mapping_parser.py (hunk 2 above)
Comment (prescod) [reply cumulusci/tasks/bulkdata/mapping_parser.py:26]: 3441a8c8

File: cumulusci/tasks/bulkdata/mapping_parser.py (hunk 1 above)
Comment (prescod) [reply cumulusci/tasks/bulkdata/mapping_parser.py]: 3441a8c8

File: cumulusci/tasks/bulkdata/utils.py (hunk 4 above)
Comment (davisagli) [reply cumulusci/tasks/bulkdata/utils.py:18]: I don't see any mention of it in our products and think it's unlikely to be used elsewhere.

Comment (davisagli) [reply cumulusci/tasks/bulkdata/utils.py]: I removed it: https://github.com/SFDO-Tooling/CumulusCI/pull/1772/commits/766f4612dcb6f99c27ce1f7a9523a4fbee5ac162

**Pull request merged**